File: cfapi.py
import requests, json, time, os, functools

import package
import util
# All of this is based on https://github.com/Mondanzo/mc-curseforge-api
BASE_URL = "https://addons-ecs.forgesvc.net/api/v2/addon"
USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:96.0) Gecko/20100101 Firefox/96.0"
REPOSITORY = "https://curseforge.com"

def add_pkg_to_tracked(cf_id: int):
    j = util.get_cf_idx()
    j['cf_ids'] = util.arr_add(j['cf_ids'], cf_id)
    # print(j)
    f = open("./assets/pkgs_to_track.json", "w")
    f.write(json.dumps(j))
    f.close()

def link_id(cf_id: int, local_id: int):
    j = util.get_cf_idx()
    for i, pkg in j['packages']:
        if(pkg['cf_id'] == cf_id):
            j['packages'][i]['local_id'] = local_id
            break
    
    f = open("./assets/pkgs_to_track.json", "w")
    f.write(json.dumps(j))

def set_last_indexed(local_id: int, last_modified):
    j = util.get_cf_idx()
    for i, pkg in j['packages']:
        if(pkg['local_id'] == local_id):
            j['packages'][i]['last_modified'] = last_modified
            break
    
    f = open("./assets/pkgs_to_track.json", "w")
    f.write(json.dumps(j))

def has_pkg_from_cf(cf_id: int) -> int:
    # print(f"has_pkg_from_cf: {cf_id}")
    j = util.get_cf_idx()['packages']
    # print(j)
    # pkgs = [pkg for pkg in j if pkg['cf_id'] == cf_id]
    pkgs = list(filter(lambda pkg: pkg['cf_id'] == cf_id, j))
    if len(pkgs) == 0:
        return -1
    return pkgs[0]['local_id']

def add_pkg_to_idx(cf_id: int):
    # first, we build the dependency tree
    
    dependency_list = {f"{cf_id}": []}
    root = cf_id
    ids_todo = [cf_id]
    resolved_ids = []
    search_for_deps = True
    while search_for_deps:
        print(dependency_list)
        if len(ids_todo) == 0:
            search_for_deps = False
            break
        
        current_id = ids_todo.pop()
        print(f"doing {current_id}")

        releases = get_releases(current_id)
        
        started_at = dependency_list
        for rel in releases:
            for dep in rel['dependencies']:
                if dep['type'] != 2:
                    if resolved_ids.count(dep['addonId']) == 0:
                        ids_todo = util.arr_add(ids_todo, dep['addonId'])
                    try:
                        dependency_list[f"{current_id}"] = util.arr_add(dependency_list[f"{current_id}"], dep['addonId'])
                    except:
                        dependency_list[f"{current_id}"] = util.arr_add(None, dep['addonId'])
                else:
                    print(f"Ignoring optional (pass 1) {dep} for {current_id}")
        
        resolved_ids = util.arr_add(resolved_ids, current_id)

    # at this point we have all the needed packages
    
    # make an array with the order of installation
    # ids_todo can be reused cuz they're empty
    left_to_resolve = dependency_list
    all_done = False
    ids_todo = [root]
    ordered = [root] # end with the root
    while not all_done:
        if len(ids_todo) == 0: # no more unresolved packages
            all_done = True
            break

        print(ordered)
        current_id = ids_todo.pop()
        
        try:
            _ = left_to_resolve[f"{current_id}"]
        except:
            left_to_resolve[f"{current_id}"] = []

        for dep in left_to_resolve[f"{current_id}"]:
            ids_todo.append(dep)
            # always move the required id to the top
            ordered = list(filter(lambda id: id != dep, ordered))
            ordered.append(dep)
    
    print(f"final order: {ordered}")

    all_added = False
    while not all_added:
        if len(ordered) == 0:
            all_added = True
            break

        current_id = ordered.pop()
        print(f"installing {current_id}")
        if has_pkg_from_cf(current_id) > -1:
            print(f"has {current_id}")
            continue

        # add the package to our index
        pkg = get_package(current_id)
        rels = get_releases(current_id)
        print(f"{len(rels)} releases")
        authors = []
        for au in pkg['authors']:
            authors.append(au['name'])

        p = package.Package.create_new(pkg['name'], pkg['slug'], pkg['summary'], authors, repository=REPOSITORY, repository_id=current_id)

        # do everything we have to with the index file
        add_pkg_to_tracked(p.repository_id)
        
        cf_idx = util.get_cf_idx()
        cf_idx['packages'].append({"cf_id": current_id, "local_id": p.id, "last_modified": util.time_ms()})
        f = open("./assets/pkgs_to_track.json", "w")
        f.write(json.dumps(cf_idx))
        f.close()


        for rel in rels:
            if len(rel['gameVersion']) == 0:
                print(f"game version not found for {rel['displayName']} id {rel['id']}. Skipping")
                continue
            deps = []
            for dep in rel['dependencies']:
                if dep['type'] != 2:
                    depid = has_pkg_from_cf(dep['addonId'])
                    deps.append( {"pkg_id": depid, "release_id": package.get_newest_release(depid, rel['gameVersion'][0])['id'], "repo": REPOSITORY})
                else:
                    print(f"Ignoring optional (pass 2) {dep} for {p.id}")
            
            r = package.Release.create_new("unknown", rel['gameVersion'][0], deps, p.id, released=util.iso_str_to_ms(rel['fileDate']), direct_link=rel['downloadUrl'].replace("https://edge.", "https://media."), prefer_link=True)

def update_releases(local_id: int):
    p = package.Package.obj_from_id(local_id)
    releases = get_releases(p.repository_id)

    
    new_releases = len(releases) - len(p.releases)
    print(f"New releases for {p.name}: {new_releases}")
    
    
    # for rel in releases:
    #     package.Release.create_new("unknown", rel['gameVersion'][0], [], p.id, direct_link=rel['downloadUrl'].replace("https://edge.", "https://media."), prefer_link=True)
    # set_last_indexed(local_id, util.time_ms())

def create_tracked_pkgs(cf_ids):
    for id in cf_ids:
        print(f"create_tracked_pkgs() {id}")
        if has_pkg_from_cf(id) > -1:
            # update_releases(has_pkg_from_cf(id))
            continue
        
        else:
            add_pkg_to_idx(id)

def crawl_curseforge(start_index: int):
    page_size = 50
    i = start_index
    while True:
        packages = get_page(i)
        for pkg in packages:
            create_tracked_pkgs([pkg['id']])

        i += 1


@functools.lru_cache(maxsize=256)
def get_package(cf_id: int):
    got_result = False
    while not got_result:
        a = requests.get(BASE_URL+ f"/{cf_id}/", headers={"User-Agent": USER_AGENT}).text
        try:
            pkg = json.loads(a)
            got_result = True
        except:
            continue
    # print(a)
    return pkg

@functools.lru_cache(maxsize=256)
def get_releases(cf_id: int):
    got_result = False
    while not got_result:
        a = requests.get(BASE_URL+ f"/{cf_id}/files/", headers={"User-Agent": USER_AGENT}).text
        try:
            releases = json.loads(a)
            got_result = True
        except:
            continue

    # print(a)
    return releases

def get_page(page: int):
    PAGE_SIZE=50

    got_result = False
    while not got_result:
        a = requests.get(BASE_URL+ f"/search?gameId=432&sectionId=6&pageSize={PAGE_SIZE}&index={page}", headers={"User-Agent": USER_AGENT}).text
        try:
            packages = json.loads(a)
            got_result = True
        except:
            continue

    print(a)
    return packages

File: src/cfapi.ts
import * as packages from "./package"
import * as util from "./util"
import * as filedef from "./filedef"

// All of this is based on https://github.com/Mondanzo/mc-curseforge-api
const BASE_URL = "https://addons-ecs.forgesvc.net/api/v2/addon"
const USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:96.0) Gecko/20100101 Firefox/96.0"
const REPOSITORY = "https://curseforge.com"

class CfRelease {
    id: number;
    displayName: string;
    fileName: string;
    fileDate: string;
    fileLength: number;
    releaseType: number;
    fileStatus: number;
    downloadUrl: string;
    isAlternate: boolean;
    alternateFieldId: number;
    dependencies: Array<{addonId: number; type: number;}>;
    isAvailable: boolean;
    modules: Array<{foldername: string; fingerprint: number;}>;
    packageFingerprint: number;
    gameVersion: Array<string>;
    installMetadata: boolean;
    serverPackFileId: boolean;
    hasInstallScript: boolean;
    gameVersionDateReleased: string;
    gameVersionFlavor: boolean;
}

class CfPackage {
    id: number;
    name: string;
    authors: Array<{name: string; url: string;}>;
    summary: string;
    slug: string;
    modLoaders: Array<string>;
    dateModified: string;
}

let cf_package_cache = new Map<number, CfPackage>();
let cf_release_cache = new Map<number, Array<CfRelease>>();

export function get_releases(cf_id: number): Array<CfRelease> {
    if (cf_release_cache.has(cf_id)) {
        return cf_release_cache.get(cf_id)
    }
    let resp = util.get_sync(`${BASE_URL}/${cf_id}/files`);
    let tmp = JSON.parse(resp);
    cf_release_cache.set(cf_id, tmp);
    return tmp;
}

export function get_package(cf_id: number): CfPackage {
    if (cf_package_cache.has(cf_id)) {
        return cf_package_cache.get(cf_id);
    }
    let resp = util.get_sync(`${BASE_URL}/${cf_id}`);
    let tmp = JSON.parse(resp);
    cf_package_cache.set(cf_id, tmp);
    return tmp;
}

export function add_pkg_to_tracked(cf_id: number, sslug: string) {

}

// TODO: Cache curseforge api responses
export function index_package(id_to_add: number) {
    // Step one: build a dependency tree
    let dependency_tree = new Map<number, Set<number>>();
    let ids_to_do = [id_to_add];
    let resolved_ids = new Set<number>();
    while (true) {
        if (ids_to_do.length == 0) {
            break;
        }
        let current_id = ids_to_do.pop();

        let releases = get_releases(current_id);

        for (const release of releases) {
            for (const dependency of release.dependencies) {


                switch(dependency.type) {
                    case 1:
                    case 2:
                    case 5:
                        break;
                    case 6:
                    case 3:
                        // required
                        if (!resolved_ids.has(dependency.addonId)) {
                            ids_to_do.push(dependency.addonId);
                        }
                        let tmp = dependency_tree.get(current_id);
                        if (tmp === undefined) {
                           tmp = new Set<number>() 
                        }
                        util.print_debug(tmp)
                        tmp.add(dependency.addonId);
                        dependency_tree.set(current_id, tmp);
                        break;
                    case 4:
                        util.print_debug(`Encountered 'Tool' (4) dependency type for ${current_id} release ${release.displayName} - skipping`);
                        break;
                    
                    default:
                        util.print_debug(`Encountered  (${dependency.type}) dependency type for ${current_id} release ${release.displayName} - skipping. Should be unreachable`);
                        break;
                }
                resolved_ids.add(dependency.addonId);
                
            }
        }
    }

    // At this point all the resolved IDs and a dependency tree
    // Get the correct order

    ids_to_do = [id_to_add];
    let ordered = [id_to_add]; // always end with the root package

    while (true) {
        if (ids_to_do.length == 0) {
            break;
        }
        util.print_debug(ordered);
        let current_id = ids_to_do.pop();
        let tmp = dependency_tree.get(current_id) ?? new Set<number>();
        for (const dependency of tmp.values()) {
            ids_to_do.push(dependency);
            ordered = ordered.filter( (id) => id != dependency);
            ordered.push(dependency);
        }
    }
    util.print_debug(`Final order: ${ordered}`);

    while (true) {
        if (ordered.length == 0) {
            break;
        }
        let current_id = ordered.pop();
        util.print_debug(`Installing ${current_id}`);

        let pkg = get_package(current_id);
        let releases = get_releases(current_id);
        util.print_debug(`${releases.length} releases`);
        let authors = pkg.authors.map( (author) => author.name);

        let p = packages.Package.create_new(pkg.name, pkg.summary, authors, [], REPOSITORY, pkg.slug)

        packages.TrackedPackage.mark_updated(pkg.id, p.slug);
        for (const release of releases) {
            let loader = "Forge";
            if (release.gameVersion.includes("Fabric")) {
                loader = "Fabric";
            }
            release.gameVersion = release.gameVersion.filter( (str) => str != "Fabric" && str != "Forge")

            if (release.gameVersion.length == 0) {
                util.print_debug(`Game version for ${release.displayName} id ${release.id} not found, skipping`);
                continue;
            }
            let dependencies = new Array<string>(); // sslugs
            for (const dependency of release.dependencies) {
                let ppkg = get_package(dependency.addonId);
                let ppkg_loc = packages.Locator.from_short_slug(`${REPOSITORY}->${ppkg.slug}->0`)
                let local_pkg = packages.locator_to_package(ppkg_loc, filedef.get_index().packages);
                let rel_id = packages.get_desired_release(local_pkg, release.gameVersion[0]).id;
                let slug = new packages.Locator(REPOSITORY, local_pkg.slug, rel_id);
                dependencies.push(slug.short_slug);
            }
            let r_id = p.next_release_id;
            let r = packages.Release.create_new(
                "unknown", 
                release.gameVersion[0],
                dependencies, 
                `${p.repository}->${p.slug}->${r_id}`, 
                Date.parse(release.fileDate), 
                release.downloadUrl, true
            );
        }

    }
    // to_add.add(id_to_add);
    // for (const cf_id of to_add) {
    //     let pkg = get_package(cf_id);
    //     // create new package, add it to tracked and index
    //     let p = packages.Package.create_new(pkg.name, pkg.summary,pkg.authors.map( (author) => {return author.name}), [], REPOSITORY, pkg.slug)

    //     packages.TrackedPackage.mark_updated(cf_id, p.slug);
        
    //     // index the releases and get all dependencies in order in which they should be installed
    //     const releases = get_releases(cf_id);
        
    //     for (const release of releases) {
    //         let deps = [];
    //         let loader = "Forge";
    //         if (release.gameVersion.includes("Fabric")) {
    //             loader = "Fabric";
    //         }
    //         release.gameVersion = release.gameVersion.filter( (str) => str != "Fabric" && str != "Forge")

              
    //     }
        
        
    // }

}